### PR TITLE
DBZ-2699 implement skipped.operation in core

### PR DESCRIPTION
**Pull request by blcksrx**
implement `skipped.operation` in core and also implement it on `Postgresql`

### CODE REVIEW

**Comment by Naros** [comment]
@blcksrx does mean the changes for DBZ-2696 (PostgreSQL) and DBZ-2697 (SQL Server) are no longer needed if you're now incorporating this change inside Debezium core?

**Comment by blcksrx** [comment]
About Postgres Im going to close it. But sqlserver Pr has more performance than this

**Comment by jpechane** [comment]
@blcksrx Thanks fro the contribution. There is still an issue as events filtered metric is not increased. Also transaction context might need to be updated as in method `io.debezium.pipeline.txmetadata.TransactionMonitor.dataEvent(DataCollectionId, OffsetContext, Object, Struct)`

**Comment by blcksrx** [comment]
@jpechane Hi! I'm sorry to answer too late, Can you describe the issues more in detail?

**Comment by jpechane** [comment]
@blcksrx Please check the top of `io.debezium.pipeline.EventDispatcher.dispatchDataChangeEvent(T, ChangeRecordEmitter)` that's the place where filtering happens and metric is updated. Hance my proposola for `getOperation()` to be be pull-up so you can extend and centralize the filtering logic there. That would solve all issues.

**Comment by rk3rn3r** [comment]
@jpechane @blcksrx don't you think when we move code more an more up we could also move out that complexity out of Debezium and into an SMT? The benefit would be less complexity inside Debezium which leads to simpler test-cases, more stability, more performance where that use-case isn't active, etc.

**Comment by jpechane** [comment]
@rk3rn3r The filtering is already present for MySQL. SMTs has their own price. If we choose the option I've proposed then the code will simplify significantly.

**Comment by rk3rn3r** [comment]
I know, and also MongoDB. But removing or better: extraction of a feature to an SMT is a valuable thing to me, at least for the cases where we are about to move the code directly in front of where we emit the Kafka Connect SourceRecord! Then the same thing can be done with an SMT and we keep the Debezium (core) code cleaner.
Having this in connectors where we for example configure the captured datastore to skip events for us is valueable. Using a connector/datastore specific feature is good then. But I am not convinced that too much core complexity is good long term, especially when we already have a layer to address such filtering.

**Comment by blcksrx** [comment]
So what should I do now? I've struggled :)

**Comment by jpechane** [comment]
@blcksrx Sorry for the confusion. Please continue with your work. Could you please address my comments if possible?

**Comment by blcksrx** [comment]
It's okay. I try my best

**Comment by blcksrx** [comment]
@jpechane I didn't get why the test fails. It's a simple code. and also I wonder why `assertInsert` does not work for it!
if you check the logs you will get what I mean. by the way, can you help me? :))

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java [hunk 1]
```diff
@@ -2381,6 +2382,44 @@ public void shouldProduceMessagesOnlyForConfiguredTables() throws Exception {
...
+        TestHelper.execute("INSERT INTO s2.a VALUES(202, 202, 'insert-202')");
+        TestHelper.execute("DELETE FROM s2.a WHERE pk=202");
+        TestHelper.execute("INSERT INTO s2.a VALUES(203, 203, 'insert-203')");
+
+        SourceRecords records = consumeRecordsByTopic(3);
+        List<SourceRecord> recordsForTopic = records.recordsForTopic(topicName("s2.a"));
+
+        assertThat(recordsForTopic.size()).isEqualTo(3);
+        assertInsert(recordsForTopic.get(0), PK_FIELD, 201);
+        assertInsert(recordsForTopic.get(1), PK_FIELD, 202);
```
Comment (rk3rn3r) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java:2411]: hey @blcksrx. I think the current failing test comes from the test that you added.

`recordsForTopic.get(1)` comes from `TestHelper.execute("UPDATE s2.a SET bb='update-201' WHERE pk=201");` statement which is an update and not a CREATE/INSERT.

**Comment by blcksrx** [comment]
Actually it should not happens because of the connector has the config and it should not bring u and d operation.

**Comment by rk3rn3r** [comment]
hey @blcksrx! I added a commit for testing if it fixes the dependency problem that occured in the debezium-testing package build

**Comment by blcksrx** [comment]
Hooray, Thanks @rk3rn3r!

**Comment by rk3rn3r** [comment]
Loool... and fu**d up the commit message. :man_shrugging:

**Comment by rk3rn3r** [comment]
Fixed the commit msg and the dependency issue seems resolved.
Now it leaves the old issue: `PostgresConnectorIT.shouldEmitNoEventsForSkippedUpdateAndDeleteOperations:2411->AbstractConnectorTest.assertInsert:814 expected:<20[2]> but was:<20[1]>`

**Comment by rk3rn3r** [comment]
Looks like my last change solved the PG connector test on the debezium-testing CI build workflow. The SQL Server issue might be unrelated. I will retrigger the build and see if it is solved then.

**Comment by blcksrx** [comment]
Thanks @rk3rn3r!  the problem has been solved

File: .github/workflows/testing-workflow.yml [hunk 2]
```diff
@@ -41,4 +41,4 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-maven-
       - name: Check changes in Debezium Testing
-        run: mvn clean install -B -pl debezium-testing -am -amd -Passembly -Dcheckstyle.skip=true -Dformat.skip=true -Drevapi.skip -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
+        run: mvn clean install -B -pl debezium-assembly-descriptors,debezium-core,debezium-testing -am -amd -Passembly -Dcheckstyle.skip=true -Dformat.skip=true -Drevapi.skip -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
```
Comment (rk3rn3r) [inline .github/workflows/testing-workflow.yml:44]: I added both dependencies as it looked like sometimes the package for `debezium-assembly-descriptors` was missing and the build terminated. And later it looked like a cached version of `debezium-core` (without the applied changes) was used and the changes were not active and tests were failing. I had no time to dig deeper into that but maybe @jpechane have a better idea. It might be maven cache / cache-key related.

**Comment by rk3rn3r** [comment]
Woop! Woop!

File: .github/workflows/testing-workflow.yml (hunk 2 above)
Comment (jpechane) [reply .github/workflows/testing-workflow.yml:44]: That was exactly the problem, it has not build/published the core module (and it is hard to guess why) and deps and either were taken from Internet or from cache with result that the code was not aligned.

Comment (rk3rn3r) [reply .github/workflows/testing-workflow.yml:44]: Don't you think we need to change the cache key then in order to not fetch a cache that has assets for, for example,  debezium-core and use outdated code?

Comment (jpechane) [reply .github/workflows/testing-workflow.yml:44]: Well, you are right, we can exclude `io/debezium` dir from caching. There was originally plan that there will be a single job that will build the artifacts and then the connector jobs will not do the build but testing but it seems such implementation is not necessary with selective connector building.

Comment (blcksrx) [reply .github/workflows/testing-workflow.yml:44]: Guys, can you tell me what is the problem? :))

Comment (jpechane) [reply .github/workflows/testing-workflow.yml:44]: Sorry, the discussion is a litlebit spread https://github.com/debezium/debezium/pull/2018#discussion_r544432542 :-)

**Comment by blcksrx** [comment]
@rk3rn3r  Hey Guys, I'm sorry to respond very late. I was migrating to Amsterdam and I was busy. By the way, can you check why CI failed?
the error is related to the  `ConditionTimeOut`  that maybe restarting would help :)

**Comment by rk3rn3r** [comment]
@blcksrx the CI error is unrelated. If you think the PR is ready we can have a final review.

**Comment by blcksrx** [comment]
It ready. Lets go!

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java [hunk 3]
```diff
@@ -2452,6 +2452,22 @@ public void testMaxLsnSelectStatementWithFalse() throws Exception {
...
+        final Configuration config = TestHelper.defaultConfig()
+                .with(SqlServerConnectorConfig.SNAPSHOT_MODE, SnapshotMode.INITIAL)
+                .with(SqlServerConnectorConfig.SKIPPED_OPERATIONS, "u,d")
+                .build();
+
+        start(SqlServerConnector.class, config);
+        assertConnectorIsRunning();
+        // Wait for snapshot completion
+        TestHelper.waitForSnapshotToBeCompleted();
+        consumeRecordsByTopic(1);
```
Comment (rk3rn3r) [inline debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java:2467]: Is something missing in this test?

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java [hunk 4]
```diff
@@ -2467,26 +2467,6 @@ public void shouldProduceMessagesOnlyForConfiguredTables() throws Exception {
...
-        Configuration.Builder configBuilder = TestHelper.defaultConfig()
-                .with(PostgresConnectorConfig.PUBLICATION_NAME, "cdc")
-                .with(PostgresConnectorConfig.PUBLICATION_AUTOCREATE_MODE, PostgresConnectorConfig.AutoCreateMode.FILTERED.getValue())
-                .with(PostgresConnectorConfig.TABLE_INCLUDE_LIST, "nonexistent.table");
-
-        start(PostgresConnector.class, configBuilder.build());
-        assertConnectorNotRunning();
-        assertTrue(logInterceptor.containsStacktraceElement("No table filters found for filtered publication cdc"));
-    }
-
```
Comment (rk3rn3r) [inline debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java:2489]: How is this test removal related?

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java [hunk 5]
```diff
@@ -2452,6 +2452,42 @@ public void testMaxLsnSelectStatementWithFalse() throws Exception {
...
+            assertThat(value.get("op")).isEqualTo(Envelope.Operation.CREATE.code());
+            assertThat(value.get("op")).isNotEqualTo(Envelope.Operation.UPDATE.code());
+            assertThat(value.get("op")).isNotEqualTo(Envelope.Operation.DELETE.code());
+        });
+
+        assertInsert(tableA.get(0), "id", 201);
+        assertInsert(tableA.get(1), "id", 202);
+        assertInsert(tableA.get(2), "id", 203);
+
+    }
```
Comment (rk3rn3r) [inline debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java:2489]: Would it be possible to add a test (that then will be executed for all connectors "auto-magically") in debezium-core? No is a valid answer, but it would be nice if that would be somehow possible.
I am not sure if it's possible to have a test case there or it always needs coupling to a connector implementation?

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java (hunk 4 above)
Comment (blcksrx) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java:2489]: I think it's related to some merge conflicts that I have. I will fix it.

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java (hunk 5 above)
Comment (blcksrx) [reply debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java:2489]: I will think about it to find a proper way

**Comment by rk3rn3r** [comment]
@blcksrx could you rebase your branch to latest master please?
Overall it looks like we are almost there. It would be nice if we can come up with a test case in debezium-core, but I can imagine that it might be too complex and requires an integration scenario, but would be good if you could check if it is somehow possible.

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java [hunk 6]
```diff
@@ -205,9 +208,11 @@ public void changeRecord(DataCollectionSchema schema,
                                              OffsetContext offset,
                                              ConnectHeaders headers)
                             throws InterruptedException {
-                        transactionMonitor.dataEvent(dataCollectionId, offset, key, value);
-                        eventListener.onEvent(dataCollectionId, offset, key, value);
-                        streamingReceiver.changeRecord(schema, operation, key, value, offset, headers);
+                        if (!skippedOperations.contains(operation)) {
```
Comment (rk3rn3r) [inline debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:211]: @blcksrx Am I right you said we can solve filtering more performant with db tools for some databases (like SQL Server for example)?
@jpechane you suggested a design that for such cases we can use the more performant solution instead. what was your idea about we make this possible? (like how to deactivate this global behavior here in core in favor of the more performant implementation in a connector implementation?

Comment (blcksrx) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:211]: Yes, you are right @jpechane suggest first implement it on the core, so this feature will be available for all the connector. after that we can design it for each connector with a better performance.

Comment (rk3rn3r) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:211]: I see!
@jpechane do we need to do anything in this PR to address that flexibility now or do we do this as bigger refactoring with the first connector specific implementation?

**Comment by blcksrx** [comment]
@rk3rn3r  I can't understand what is wrong, do you have any idea?

**Comment by rk3rn3r** [comment]
@blcksrx maybe you rebased at a bad state of the upstream master branch? there was some work on transaction handling and a later test fix. either rebase to master again and see if it solves the issue or try to identify the problems locally (run `mvn -am clean install`).

**Comment by blcksrx** [comment]
@rk3rn3r Finally fixed! although there is only one error on the `Postgres` container that's weird. It's unrelated to the PR and also it passes on my local.

**Review by rk3rn3r — APPROVED** [review]
LGTM! @jpechane should have another look too and review my comment above.

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java (hunk 6 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:218]: @rk3rn3r Nothing specific needed right now. MySQL is already rewritten to use the new framework and IIRC the skipped ops are already implemente dthere so it might be a good starting point. @blcksrx Could you please raise a follow-up Jira for that?

@blcksrx Also I've one more tip for micro-optimization.  There should be a boolean flag attribute `neverSkip` that is set at the start depending on whether the map is empty or not.

The condition would then look like
```
if (neverSkip ||  !skippedOperations.contains(operation))
```
This would be more performant in case skipped ops are not set.

**Comment by jpechane** [comment]
@blcksrx Thanks for the PR, sorr that it dragged so long. I've just left a comment for a small microoptimization.

Unfortunately we are at the very late stage before `1.5.0.Final` release so only bugfixes are now merged. This PR will be merged immediatelly after the 1.5.0.Final is out.

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java (hunk 6 above)
Comment (blcksrx) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:218]: @jpechane thanks for answering. I will do the optimization tonight and about another issue, currently, the `MongoDB connectors try to filter that on the `OpLog` side. I can create tickets for other connectors

**Comment by jpechane** [comment]
@blcksrx And applied, thanks a lot for your patience!

**Comment by blcksrx** [comment]
@jpechane Thanks a lot. I will continue my work to optimize this option for each connector

**Comment by gunnarmorling** [comment]
Thx! I've pushed one more commit (https://github.com/debezium/debezium/commit/35c9e846c5be520bd473d9d0d7a56aa5509ca6fc), as we shouldn't speak about "oplog" in connector docs other than MongoDB.

**Comment by blcksrx** [comment]
my bad, sorry
